python/pyspark/ml/base.py

```diff
@@ -116,3 +121,44 @@ class Model(Transformer):
     """
 
     __metaclass__ = ABCMeta
+
+
+@inherit_doc
+class UnaryTransformer(HasInputCol, HasOutputCol, Transformer):
+
+    @abstractmethod
+    def createTransformFunc(self):
+        """
+        Creates the transoform function using the given param map.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def outputDataType(self):
+        """
+        Returns the data type of the output column as a sql type
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def validateInputType(self, inputType):
+        """
+        Validates the input type. Throws an exception if it is invalid.
+        """
+        raise NotImplementedError()
+
+    def transformSchema(self, schema):
+        inputType = schema[self.getInputCol()].dataType
+        self.validateInputType(inputType)
+        if self.getOutputCol() in schema.names:
+            raise ValueError("Output column %s already exists." % self.getOutputCol())
+        outputFields = copy.copy(schema.fields)
+        outputFields.append(StructField(self.getOutputCol(),
+                                        self.outputDataType(),
+                                        nullable=False))
+        return StructType(outputFields)
+
+    def transform(self, dataset, paramMap=None):
+        transformSchema(dataset.schema())
```

Comment:
Here seems exist some problem.
The transform provide `paramMap`, but `createTransformFunc` has no way to get the passed in `paramMap`, here lost something I think.
Because custom UnaryTransformer will only need to override the `createTransformFunc`, the base class need to handle the passed in `paramMap` properly.